bugbot/bug/analyzer.py

```diff
@@ -0,0 +1,234 @@
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+
+from functools import cached_property
+from typing import Any, Iterable, NamedTuple
+
+from libmozdata import versions as lmdversions
+from libmozdata.bugzilla import Bugzilla
+
+from bugbot import utils
+
+
+class VersionStatus(NamedTuple):
+    """A representation of a version status flag"""
+
+    channel: str
+    version: int
+    status: str
+
+    @property
+    def flag(self) -> str:
+        return utils.get_flag(self.version, "status", self.channel)
+
+
+class BugAnalyzer:
+    """A class to analyze a bug"""
+
+    def __init__(self, bug: dict, store: "BugsStore"):
+        """Constructor
+
+        Args:
+            bug: The bug to analyze
+            store: The store of bugs
+        """
+        self._bug = bug
+        self._store = store
+
+    @property
+    def regressed_by_bugs(self) -> list["BugAnalyzer"]:
+        """The bugs that regressed the bug."""
+        return [
+            self._store.get_bug_by_id(bug_id) for bug_id in self._bug["regressed_by"]
+        ]
+
+    @property
+    def oldest_fixed_firefox_version(self) -> int | None:
+        """The oldest version of Firefox that was fixed by this bug."""
+        fixed_versions = sorted(
+            int(key[len("cf_status_firefox") :])
+            for key, value in self._bug.items()
+            if key.startswith("cf_status_firefox")
+            and "esr" not in key
+            and value in ("fixed", "verified")
+        )
+
+        if not fixed_versions:
+            return None
+
+        return fixed_versions[0]
+
+    @property
+    def latest_firefox_version_status(self) -> str | None:
+        """The version status for the latest version of Firefox.
+
+        The latest version is the highest version number that has a status flag
+        set (not `---`).
+        """
+        versions_status = sorted(
+            (int(key[len("cf_status_firefox") :]), value)
+            for key, value in self._bug.items()
+            if value != "---"
+            and key.startswith("cf_status_firefox")
+            and "esr" not in key
+        )
+
+        if not versions_status:
+            return None
+
+        return versions_status[-1][1]
+
+    def get_field(self, field: str) -> Any:
+        """Get a field value from the bug.
+
+        Args:
+            field: The field name.
+
+        Returns:
+            The field value. If the field is not found, `None` is returned.
+        """
+        return self._bug.get(field)
+
+    def detect_version_status_updates(self) -> list[VersionStatus] | None:
+        """Detect the status for the version flags that should to be updated.
+
+        The status of the version flags is determined by the status of the
+        regressor bug.
+
+        Returns:
+            A list of `VersionStatus` objects.
+        """
+        if len(self._bug["regressed_by"]) > 1:
+            # Currently only bugs with one regressor are supported
```

Comment:
Could you file a follow-up to expand this support? When the regressors are all fixed in the same version, it's obvious. When they are fixed in different versions, we could set the latest as "affected" and the earliest as "?"